Updates password/token_auth hashing

mneudert opened the pull request:
Refs #5728:
- Updates password hashing
- Updates token_auth hashing

work in progress...

mneudert committed:
converts user.password to VARCHAR(255)

mneudert committed:
re-hashes existing passwords with bcrypt

mneudert committed:
re-hashes legacy password after first login

mneudert committed:
converts user.token_auth to CHAR(64)

mneudert commented:
I dropped most of the travis testing work to not have it completely run all the time. That is somewhat a development artifact and of course will be removed before a merge is considered. If the actual pull request should do the full testing cycle I can easily remove it and just keep it for development in different branches (as to not completely mess up notifications/comments here).

-----

For the mentioned caveats/quirks:

Password is wrapped with bcrypt during update and an option called
`${login}_legacy_password` takes care of later identifying those migrated
users.

Upon regular login this flag triggers an automatic update. That is, password
gets re-hashed with pure bcrypt and option removed.

### password_hash configuration

Currently only the default bcrypt configuration is used. This should of course
move to some config flags. Including a change of the actual hashing algorithm
to whatever `password_hash` supports.

### password_hashed database field

As the login mechanisms allow passing an MD5 hashed password (like the
python log importer) there is a secondary storage under `password_hashed`
that wraps the MD5 password with a layer of bcrypt.

Upon every password change this field also gets a new round of hashing.

The only way I know to work around this double storage is to completely wrap
the password with bcrypt around MD5. That would also allow us to re-hash the
password during url authentication.

### testing

As the passwords are now non-deterministic most of the XML fixtures make the
tests fail. The fixture check could replace the password fields with some value
to prevent these failures. However that would mean any test using a fixture to
check a change in the password would never detect any failure...

### token_auth

Only changed the database field to a length viable to accomodate a SHA256 hash.

Every new auth token would directly be generated with that length while the old
ones would stay in place and working.

Regenerating that one single token could be done using a simple button in the
admin interface. Imho the first password change while still having an old token
should also regenerate the hash with a new one and then drop that connection.

tsteur commented on core/Updates/3.0.0-b2.php:
```diff
@@ -21,4 +23,84 @@ public static function isMajorUpdate()
+
+            $db->query(
+                'UPDATE `' . $userTable . '`'
+                . ' SET `password` = :password, `password_hashed` = :password_hashed'
```
Will there be also an update for `token_auth = hash('sha256, token_auth)`? `sha2` was only added recently in MySQL 5.5 so we might not be able to use it actually. Will need to do it in PHP unfortunately

tsteur replied:
Actually, I understand the problem is then we can no longer show the token in the UI so we can do this in a separate PR 👍

tsteur commented on core/Updates/3.0.0-b2.php:
```diff
@@ -21,4 +23,84 @@ public static function isMajorUpdate()
+    private function getUserDatabaseMigrations($queries)
+    {
+        $queries[] = $this->migration->db->changeColumn($this->userTable, 'password', 'password', 'VARCHAR(255) NOT NULL');
+        $queries[] = $this->migration->db->addColumn($this->userTable, 'password_hashed', 'VARCHAR(255) NOT NULL', 'password');
```
I'm not quite sure, do we need the `password_hashed` actually? Or could we just upgrade all `password` fields and then try 1) `password_verify($password)`, and if it fails try `password_verify(md5(password))`  or as we know by the option table which ones were upgraded we would know directly whether to apply md5 or not?

mneudert replied:
I would love to remove that field. The problem is that this would break at least the `UsersManager.getTokenAuth` API used in `piwik-log-analytics` script ([line 919 onwards](https://github.com/piwik/piwik-log-analytics/blob/master/import_logs.py#L919)) when using the `--password` parameter.

It being used during testing is probably easily changed. The only occurrence I found was in the `ManySitesImportedLogs` fixture.

But there may be other things that rely on calling `UsersManager.getTokenAuth` so to me it looks like it might be a necessary evil for the time being.

tsteur replied:
So my thoughts on this: Can we maybe change `getTokenAuth` to return the current `auth_token` for a user? This may be the hashed `sha256(token_auth)` version or as long as it is still saved as md5 the actual token. The method would try to log in the user and if username / password combination is correct return that token.

One problem might be that we are expecting an md5 password there whereas new passwords are stored as `password_hash($rawPassword)`. So now I am thinking would it be so bad to instead always save passwords as `password_hash(md5($rawPassword))`. So all user passwords would be immediately hashed, no option entry is needed and we stay backwards compatible with that `getTokenAuth` etc. I think that might make everything more easy

tsteur replied:
BTW: We would then need a new method `generateTokenAuth` but this would be likely not needed in the API :) `getTokenAuth()` would then do what it actually says meaning returning a token, not generating it :)

mneudert replied:
I think `getTokenAuth` and `generateTokenAuth` require the separation you mentioned already for being able to include any random data. Before accessing the token the authentication should have already happened so it should be perfectly fine to just access the user record and return what is found there.

The generation API might still need some sort of public interface as with random generation there is no need to reference the password anymore. So the token could be generated completely independent from password changes. The first password change should automatically do the migration away from the old MD5 token with an updated warning in the interface.

Regarding the passwords I also thought about just applying bcrypt around already hashed MD5 passwords. Besides rendering the legacy option obsolete it would ensure full backwards compatibility without the current "hack".

It leaves some of the problems with MD5 in place like collision vulnerability or the fact that bcrypt now works on a fixed length hash instead of potential 72 chars. But that is already a problem due to the API requirements for the MD5 login. At least it made the problem obvious :)

tsteur replied:
+1 to no longer regenerate token when password changes but then we need to have a feature in the UI to generate a new token. I'd say we could just create a new issue for this to not generate a new token when password changes to keep things easy and doable for now. I'd still include the username in the generation but password not needed anymore indeed.

> Regarding the passwords I also thought about just applying bcrypt around already hashed MD5 passwords. 
I really like that idea. In Piwik 4 we could then still think about removing the md5($password) and breaking the API.

tsteur replied:
Will you work on this in the next days so we can get rid of `password_hashed` column and maybe merge soon? :)

tsteur commented on plugins/Login/Auth.php:
```diff
@@ -60,11 +64,43 @@ public function authenticate()
+            return new AuthResult(AuthResult::FAILURE, $login, null);
+        }
+
+        $legacyOption = sprintf('%s_legacy_password', $user['login']);
```
I wonder if we need the option or whether we could always try `password_verify($password, $user['password']) || password_verify(md5($password), $user['password'])` . It may be slightly less secure as it could match when someone enters a wrong password but the chances for this are quite unlikely. Probably it is better to have the option table 👍

tsteur commented:
Looks promising so far 👍  What would be nice to have in this PR already is to generate the auth token randomly instead of based on username + password. I can't point you to where it is done right now as I cannot run my IDE right now but I'm sure you'll find it. Let me know if not

mneudert commented:
The token hashing is something I am working on locally. I will include it here once I have something working well enough to receive a look.

Just changing it (`plugins/UsersManager/API.php` should be the place) looks promising, but the calls are quite scattered. Running the tests is taking quite some time to find everything :D

tsteur commented:
BTW: With token hash generation I mean https://github.com/piwik/piwik/blob/2.17.0/plugins/UsersManager/API.php#L793

Here it may be fine to do just something like `return md5($userLogin . $md5Password . microtime() . \Piwik\Common::generateUniqId() . Config::getInstance()->General['salt'])` instead for now :)

mneudert committed:
creates/updates users with randomized token_auth

mneudert committed:
performs authentication for getTokenAuth

tsteur commented:
@mneudert  are you done with your changes so far? I'd have a look at the weekend and give it a test / review as we want to release Piwik 3 in the near future :)

mneudert commented:
It is missing the "click on a button and renew your token" UI-Stuff. However everything non-visible should be at least ready for review.

mneudert committed:
hides password change confirmation for modern token_auth

mneudert committed:
shows authentication token on personal settings page

mneudert committed:
adds hint about token regeneration to api page

mneudert committed:
allows users to regenerate their token_auth

mneudert committed:
allows admins to regenerate users token_auth

tsteur commented:
@mattab ok to have the UI feature to re-generate token at any time and to no longer show the token in the UI? This is pretty much standard behaviour nowadays on all secure platforms / services. This way we get secure tokens and secure passwords in DB finally while existing tokens would still work. We would just need to update FAQs etc. @mneudert did you solve the problem that we send along the auth token in the UI when issueing API requests via XHR? Or did I understand it wrong and tokens are pretty much still stored in plain text?

mneudert commented:
The tokens are still stored plaintext but disconnected from the passwords. Having temporary tokens for UI (and external tools like the python-log-importer) would of course be nice, but I think that is more fit for a separate PR working only on that part.

tsteur commented:
:+1: Awesome, I hope to have a look today

tsteur commented on core/Updates/3.0.0-b4.php:
```diff
@@ -0,0 +1,98 @@
+
+        foreach ($users as $user) {
+            $db->query(
+                'UPDATE `' . $userTable . '`'
```
I'm not sure if we should show those update statements in the UI @mattab ? There shouldn't be any sensitive password leak

@mneudert we could add them to the migrations via `$this->migration->db->boundSql($query, $bind)`

FYI: Usually via use `?` instead of `:password`. Not sure if there would be any problems with any DB adapters or so. Maybe better to use `?` instead.

mneudert replied:
:white_check_mark: (marking for myself before updating the PR)

If the question mark is the usual style we should stick with it. No need to risk any DB adapter to go haywire.

I decided to not include them in the migration listing thinking about the "query per user" that would appear. It might not be a problem at all but it might just make the migration log completely unusable when having too many users.

The only update I can find using the direct query style is `2.10.0-b5'. Should the updater take care of the queries even if they won't get shown for compatibility/style reasons here?

tsteur replied:
@mattab do you have thoughts on whether we need to show the queries? Probably we don't need to show them as they would be executed afterwards anyway when they perform the update so probably better to maybe not show them.

Updater will take care of them anyway 👍

mattab replied:
Agreed it is not needed to display them during updater in this case

tsteur commented:
Note to myself: We need to add a note to developer changelog that after this update there is no "going back" possible. This can be tricky when switching between branches regularly. In my case on my local machine I made a copy of the user table to be able to switch between branches more easily in the  future. We might need a command for this in the future or so. Any ideas appreciated @sgiehl @mattab

tsteur commented on plugins/UsersManager/angularjs/personal-settings/personal-settings.controller.js:
```diff
@@ -7,13 +7,13 @@
     angular.module('piwikApp').controller('PersonalSettingsController', PersonalSettingsController);
 
-    PersonalSettingsController.$inject = ['piwikApi'];
+    PersonalSettingsController.$inject = ['piwikApi', 'window'];
```
Typo: should be `$window`. Causes the personal settings to not be shown

mneudert replied:
:white_check_mark: (marking for myself before updating the PR)

tsteur commented on plugins/UsersManager/angularjs/manage-users/manage-users.controller.js:
```diff
@@ -176,6 +180,23 @@
         };
 
+        this.regenerateUserTokenAuth = function (userLogin) {
+            console.log('foo');
```
console.log should be removed

mneudert replied:
:white_check_mark: (marking for myself before updating the PR)

tsteur commented on plugins/UsersManager/templates/userSettings.twig:
```diff
@@ -107,6 +107,15 @@
+
+    <button class="btn btn-link"
+            ng-controller="PersonalSettingsController as personalSettings"
+            ng-click="personalSettings.regenerateTokenAuth()">{{ 'UsersManager_RegenerateTitle'|translate }}</button>
```
Maybe we could also add something like "Once the token has been regenerated don't forget to update the token in case you are using the token to export reports via the reporting API, to track data or to embed widgets"

tsteur replied:
I have no clue honestly, I don't mind people being logged out we would only need to mention it

mneudert replied:
:white_check_mark: (marking for myself before updating the PR)

tsteur commented on plugins/UsersManager/templates/index.twig:
```diff
@@ -178,6 +178,7 @@
                 {% if showLastSeen is defined and showLastSeen %}
                 <th>{{ 'UsersManager_LastSeen'|translate }}</th>
                 {% endif %}
+                <th>token_auth</th>
```
Can we maybe remove this column and instead use an icon in front of the shown token? The button is quite big in there and rarely needed. eg 

`<span class="icon-reload"></span> token...` We could maybe also find a new icon and add it

mneudert replied:
Yeah, that part is a rather ugly one. Personally I would like to completely separate the token management from the user management, like just having a separate `piwik-content-block` only listing the tokens and the regenerate buttons next to them.

The warning message for passwords should not feel to confusing having separate blocks. From a responsive layout perspective it might also be easier on the eyes to have those long strings separated if there are any plans to go like below 540px for everything (currently the users block needs like 9xx pixels and displaying the data veritcally (each cell as a row) might be nicer with less data per user).

tsteur replied:
We could have a separate content block for it but it could be a long list when there are > 30k users. We would need to have a user select field or a user login auto complete field similar as we have for "give view access" feature under "Manage access". It is only visible to users with admin access but not to users with super user access. We could later reuse the same logic from "give view access" .

The whole users manager needs a big re-design eventually. For now it would be good to just move that big button out and maybe replace it with a tiny icon in the existing column. In another PR we could add a new content block where users can type in the login name or email address and then re-generate their token

mneudert replied:
:white_check_mark: (marking for myself before updating the PR)

Using`.icon-reload` seems to fit nicely (displaying it already before token expansion).

tsteur commented on plugins/UsersManager/API.php:
```diff
@@ -513,6 +514,22 @@ public function getUsersHavingSuperUserAccess()
     }
 
     /**
+     * Regenerate the token_auth associated with a user.
```
We should add a comment that the same auth token cannot be used afterwards if $userLogin is the own login.

mneudert replied:
Is there any way to grab the authentication and force a re-authentication with the regenerated token (only if user matches the changed one)? Just like the JS side it might be a good thing to not bother people with too many logouts even if regenerating tokens should be a rare action to take.

tsteur replied:
> grab the authentication and force a re-authentication with the regenerated token

I'm not quite sure what you mean. It is a rare action so pretty much anything should be fine I'd say :)

mneudert replied:
:white_check_mark: (marking for myself before updating the PR)

tsteur commented on plugins/UsersManager/API.php:
```diff
@@ -513,6 +514,22 @@ public function getUsersHavingSuperUserAccess()
+    public function regenerateTokenAuth($userLogin)
+    {
+        Piwik::checkUserHasSuperUserAccessOrIsTheUser($userLogin);
+
```
We need to add a `$this->checkUserIsNotAnonymous($userLogin);`

mneudert replied:
:white_check_mark: (marking for myself before updating the PR)

tsteur commented on plugins/UsersManager/API.php:
```diff
@@ -513,6 +514,22 @@ public function getUsersHavingSuperUserAccess()
+    /**
      * Updates a user in the database.
      * Only login and password are required (case when we update the password).
      * When the password changes, the key token for this user will change, which could break
```
We can remove this comment now, and maybe mention that the token will only change if a legacy token is being used

tsteur commented on plugins/UsersManager/API.php:
```diff
@@ -780,16 +805,41 @@ private function isUserTheOnlyUserHavingSuperUserAccess($userLogin)
+            return $user['token_auth'];
+        }
+
+        throw new Exception(Piwik::translate('UsersManager_ExceptionTokenAuthCredentials', $userLogin));
```
I'm not 100% sure about this. I think we should maybe in this case still return a random token as existing applications may depend on this behaviour. Eg in Piwik Mobile we call this method, and then check whether the returned token is valid by calling `SitesManager.getSitesWithViewAccess`. We would need to test if we break Piwik Mobile by throwing an exception here. Would it be possible to simple generate a random token here and return it? It should not be based on the Piwik salt that is used in `createTokenAuth` as one could otherwise over time guess the Piwik salt. Maybe just `return hash(
            'sha256',
            $userLogin . Common::generateUniqId())
        );`

mattab replied:
:+1: `UsersManager.getTokenAuth` is one of these APIs that many users rely on, and it would be great from a API user experience perspective to maintain BC

mneudert replied:
:white_check_mark: (marking typo for myself before updating the PR)

The exception might of course be a bit rough so no problems with changing that. Even with the old logic the token would be invalid without matching the proper username/password combination.

tsteur commented on plugins/UsersManager/Controller.php:
```diff
@@ -458,6 +459,31 @@ public function recordUserSettings()
+     * Regenerates a users API authentication token.
+     * @throws Exception
+     */
+    public function regenerateTokenAuth()
```
Is it instead possible to directly call `UsersManager.regenerateTokenAuth` API from JavaSCript?

mneudert replied:
:white_check_mark: (marking typo for myself before updating the PR)

tsteur commented on plugins/UsersManager/Model.php:
```diff
@@ -241,10 +242,17 @@ public function getUsersHavingSuperUserAccess()
-             'alias'      => $alias,
-             'email'      => $email,
-             'token_auth' => $tokenAuth
+            'password'   => password_hash($password, PASSWORD_BCRYPT),
```
We have `password_hash(..., PASSWORD_BCRYPT)` quite often. Can we maybe put it in a static method somewhere or does it maybe rather not make sense? Like `Auth::hashPassword($password)`. Maybe same for other rehash methods etc?

mneudert replied:
:white_check_mark: (marking for myself before updating the PR)

tsteur commented on plugins/UsersManager/angularjs/manage-users/manage-users.controller.js:
```diff
@@ -127,7 +127,11 @@
+                    var password = $('tr#' + idRow).find('input#password').val();
+                    var token    = $('tr#' + idRow).find('td#token_auth_user').data('token') || '';
+
+                    if ('-' !== password && 32 === token.length()) {
```
Should this be `token.length`?

mneudert replied:
:white_check_mark: (marking for myself before updating the PR)

tsteur commented on plugins/UsersManager/angularjs/personal-settings/personal-settings.controller.js:
```diff
@@ -34,8 +34,26 @@
+
+            piwikApi.withTokenInUrl();
+            piwikApi.post({
+                module: 'UsersManager', action: 'regenerateTokenAuth', format: 'json'
```
Here we could call the API directly as well maybe just like in sites manager? Also mentioned this in `Controller.php` so we could get rid of that controller method

mneudert replied:
:white_check_mark: (marking typo for myself before updating the PR)

Was already done in `manage_users.controller.js`. So the controller was not needed in the first place.

tsteur commented on plugins/UsersManager/angularjs/personal-settings/personal-settings.controller.js:
```diff
@@ -49,7 +67,7 @@
             }
 
-            if (this.password) {
+            if (this.password && 32 === tokenAuth.length()) {
```
Should be probably `tokenAuth.length` as well. `.length()` results in an error

mneudert replied:
:white_check_mark: (marking for myself before updating the PR)

tsteur commented on plugins/UsersManager/templates/userSettings.twig:
```diff
@@ -107,6 +107,15 @@
 
+<div piwik-content-block
+     content-title="{{ 'UsersManager_TokenAuth'|translate|e('html_attr') }}">
+    <p id="token_auth_user" data-token="{{ userTokenAuth }}">{{ userTokenAuth }}</p>
```
Do we need the token in the attribute again or could we just read `$('#token_auth_user').text()`? Not important though I think

tsteur commented:
Looks promising, great work 👍  Mentioned a few things but they are probably "easy" to change

mattab commented:
>  ok to have the UI feature to re-generate token at any time and to no longer show the token in the UI? This is pretty much standard behaviour nowadays on all secure platforms / services.

@tsteur Interesting point. In the "Manage users" listing we only show the first few characters of each token_auth, and the full token on click (to avoid people over-looking or even screenshots leaking tokens). In the But the token_auth is also displayed in other places for example "API" page we list the full token_auth which is not consistent, we should also list only first chars. (but still leave it visible if needed as it is transmitted in the page for other reasons anyway). 

I guess what you mean is to make Piwik more secure by also completely remove token_auth from the page html output: would prevent stealing passwords over HTTP. but this would be extra work as the token_auth is used in the JS a few times, likely out of scope for this PR?

tsteur commented:
It is actually not implement yet in this PR and we need to implement this next. Either for 3.0 or for 3.X (or at the very latest 4.0). We can never show the token anywhere in the UI. Only after the token generation or re-generation. Otherwise we store the token in plain text and this is always insecure.

tsteur commented:
For JS requests etc we need to create temporary tokens or something. But this will be all a separate PR

mattab commented on plugins/API/lang/en.json:
```diff
@@ -1,5 +1,6 @@
 {
     "API": {
+        "ChangeTokenHint": "If you want to change this token, please go to your personal settings page.",
```
* would be useful to add a link to the settings page: `href="index.php?module=UsersManager&action=userSettings"`

mneudert replied:
:white_check_mark: (marking for myself before updating the PR)

mattab commented on plugins/UsersManager/API.php:
```diff
@@ -525,13 +542,22 @@ public function updateUser($userLogin, $password = false, $email = false, $alias
+
+            if (32 == strlen($token_auth)) {
+                // only refresh legacy tokens
+                $token_auth = $this->createTokenAuth($userLogin);
```
fantastic!

tsteur commented on plugins/UsersManager/API.php:
```diff
@@ -780,16 +805,41 @@ private function isUserTheOnlyUserHavingSuperUserAccess($userLogin)
+
+        if (!empty($user) && password_verify($md5Password, $user['password'])) {
+            if (password_needs_rehash($user['password'], PASSWORD_BCRYPT)) {
+                $this->model->updateUser($login, $md5Password, $user['alias'], $user['email'], $user['token_auth']);
```
sha256 is only used for tokens. Proper password_hash is used for passwords and done in updateUser

mneudert replied:
:white_check_mark: (marking typo for myself before updating the PR)

mattab commented on plugins/UsersManager/lang/en.json:
```diff
@@ -8,7 +8,7 @@
         "ApplyToAllWebsites": "Apply to all websites",
         "ChangeAllConfirm": "Are you sure you want to give '%s' access to all the websites?",
-        "ChangePasswordConfirm": "Changing the password will also change the user's token_auth. Do you really want to continue?",
+        "ChangePasswordConfirm": "Changing the password will also change the user's old token_auth. This will only happen once. Do you really want to continue?",
```
could we display this modal confirmation only when the old token will be actually changed (ie. the first time a user changes password)? 

(displaying it all the time would be bit confusing + the message mentioned "This will only happen once.")

mneudert replied:
:white_check_mark: (marking for myself before updating the PR)

mneudert replied:
I just noticed again this very same message is also displayed when changing password on the personal settings page. Would it be viable to reword it there a bit? Saying "the users" when addressing the user actually affected seems somewhat off... Perhaps just replacing the addressing part using `%s` and siwtching to "your" on that page.

mattab replied:
> Perhaps just replacing the addressing part using %s and siwtching to "your" on that page.

:+1:

mneudert replied:
Actually went with a separate message here. This allowed me to sneak a "may require you to login again" into the message while keeping it safe regarding the translation to other languages.

tsteur commented:
@mneudert will you have a chance to work on it this week / weekend? I would otherwise try to finish it on Friday or so

tsteur commented:
Actually, would you be able to work on it by Friday? I can otherwise have a look at it as some of the changes need to be quickly done. Because this change is quite big we want to get the change out soon before 3.0 RC to have enough people testing it. No worries at all if you don't find the time

mneudert commented on plugins/UsersManager/lang/en.json:
```diff
@@ -31,6 +31,7 @@
         "ExceptionPasswordMD5HashExpected": "UsersManager.getTokenAuth is expecting a MD5-hashed password (32 chars long string). Please call the md5() function on the password before calling this method.",
         "ExceptionRemoveSuperUserAccessOnlySuperUser": "Removing the Super User access from user '%s' is not possible.",
         "ExceptionSuperUserAccess": "This user has Super User access and has already permission to access and modify all websites in Piwik. You may remove the Super User access from this user and try again.",
+        "ExceptionTokenAuthCredentials": "Wrong Username and password combination.",
```
:white_check_mark: (marking for myself before updating the PR)

mneudert committed:
cleans up unused "use" statements

mneudert committed:
extracts password interaction to separate class

mneudert committed:
improves "password changes token" confirmation wording

mneudert commented:
I think I have addressed all comments (at least I marked them for myself being addressed). Reaching a "nearly mergeable" state I also removed my temporary modifications to the travis configuration. Should all be fine for another review :+1:

tsteur commented:
automated tests on travis seem to pretty much work 👍  Only few random test failures because token/password is now random but that's fine. We can mark those fields to be ignored. Gonna have a look at PR again

tsteur commented on core/Auth/Password.php:
```diff
@@ -0,0 +1,56 @@
+<?php
```
FYI: We need to set Version to 3.0.4-b4 in core/Version.php . We usually don't do this in PRs already but I think in this case it is a good idea so the update gets executed in the tests etc.

mneudert replied:
Done in a separate commit :white_check_mark:

tsteur commented on core/Updates/3.0.0-b4.php:
```diff
@@ -0,0 +1,102 @@
+                . ' SET `password` = ?'
+                . ' WHERE `login` = ?',
+                [
+                    $user['login'],
```
Shouldn't the bind order be the other way around? This also explains why my update just before did not work because no login was updated :)

mneudert replied:
:white_check_mark: Yeah, of course it should...

tsteur commented on core/Updates/3.0.0-b4.php:
```diff
@@ -0,0 +1,102 @@
+{
+    public static function isMajorUpdate()
+    {
+        return true;
```
It should not be major I think, it shouldn't take too long to update

mneudert replied:
:white_check_mark:

tsteur commented on plugins/UsersManager/templates/index.twig:
```diff
@@ -191,7 +196,13 @@
-                        <td id="token_auth_user" class="token_auth" data-token="{{ user.token_auth }}">{{ user.token_auth|slice(0, 8) }}…</td>
+                        <td id="token_auth_user" class="token_auth" data-token="{{ user.token_auth }}">
+                            <span class="token_auth_content">{{ user.token_auth|slice(0, 8) }}…</span>
+                            <button ng-click='manageUsers.regenerateUserTokenAuth({{ user.login|json_encode }}); $event.stopPropagation();'
```
Can we should a little popup forcing users to confirm this action? It would be similar to `div id= confirmPasswordChangeSelf ` only with a different message. And then you can show it before regenerating the token like  

```js
piwik.helper.modalConfirm('#confirmPasswordChange', {yes: function () {
                   // really regenerate it
                }});
```

mneudert replied:
✅

tsteur commented on plugins/UsersManager/API.php:
```diff
@@ -790,6 +844,16 @@ public function getTokenAuth($userLogin, $md5Password)
+        $user = $this->model->getUser($userLogin);
+
+        if (!$this->password->verify($md5Password, $user['password'])) {
+            return hash('sha256', $userLogin . microtime(true) . Common::generateUniqId());
```
Can we maybe move the `hash('sha256'...` into `Password` class as well? I know it doesn't make much sense to put there from the naming perspective cause Password != Token but be nice to have it always in one place :)

tsteur replied:
forget it, it's actually not needed in this case as it is only to keep BC etc. The regular logic is already in a method :)

tsteur commented on plugins/UsersManager/API.php:
```diff
@@ -790,6 +844,16 @@ public function getTokenAuth($userLogin, $md5Password)
+        }
+
+        if ($this->password->needsRehash($user['password'])) {
+            $this->updateUser($uesrLogin, $md5Password);
```
Typo, should be `$userLogin`

mneudert replied:
:white_check_mark: :roll_eyes:

tsteur commented on plugins/UsersManager/Model.php:
```diff
@@ -187,14 +194,15 @@ public function addUser($userLogin, $passwordTransformed, $email, $alias, $token
         $user = array(
             'login'            => $userLogin,
-            'password'         => $passwordTransformed,
+            'password'         => $this->passwordHelper->hash($passwordTransformed),
```
Ideally the password hashing was done outside the model but it's not too important for now (the class should be rather without logic which also makes it easier testable) but in this case it should be fine and we can always change it later

tsteur commented on plugins/UsersManager/angularjs/manage-users/manage-users.controller.js:
```diff
@@ -176,15 +186,32 @@
         };
 
+        this.regenerateUserTokenAuth = function (userLogin) {
+            var parameters = { userLogin: userLogin };
```
The manage users integration is nice 👍  Be only good to have here a confirm/alert message as well similar to in "personal settings". Just in case people click on it accidentally etc.

tsteur commented on plugins/UsersManager/tests/System/expected/test___UsersManager.getUser_login1_when_superuseraccess.xml:
```diff
@@ -2,7 +2,7 @@
 	<row>
 		<login>login1</login>
-		<password>5f4dcc3b5aa765d61d8327deb882cf99</password>
+		<password>$2y$10$/Lz0vEwEOnFNt17Gak7Js./9OJXOKN7lT5ObczOTBNtmU/ATvoGfe</password>
```
FYI: In `plugins/UsersManager/tests/System/ApiTest.php:42` we can add `password` to `xmlFieldsToRemove` array and then it won't appear anymore and they won't fail randomly

mneudert replied:
✅ That was the thing I was looking for.

tsteur commented:
Nice 👍  we are getting there 👍  A few minor tweaks missing only I'd say. Maybe @mattab can have a look as well, not sure what else we need to test but worked for me after fixing the update

mneudert committed:
bumps version to 3.0.0-b4

mneudert commented on core/Updates/3.0.0-b4.php:
```diff
@@ -0,0 +1,102 @@
+                . ' WHERE `login` = ?',
+                [
+                    $user['login'],
+                    password_hash($user['password'], PASSWORD_BCRYPT),
```
It might however be interesting to switch once a full configuration of the hashing method is available. Thinking twice that might make it worth the "effort" to directly switch to the Password class.

mneudert committed:
requires confirmation before token regeneration

mneudert commented on plugins/UsersManager/angularjs/personal-settings/personal-settings.controller.js:
```diff
@@ -34,8 +34,28 @@
 
+        this.regenerateTokenAuth = function () {
+            var parameters = { userLogin: piwik.userLogin };
+
```
✅

mneudert committed:
cleans up testing environment (passwords / tokens)

mneudert committed:
removes Password class from UsersManager model

tsteur commented on plugins/UsersManager/API.php:
```diff
@@ -780,7 +819,23 @@ private function isUserTheOnlyUserHavingSuperUserAccess($userLogin)
+     */
+    public function createTokenAuth($userLogin)
+    {
+        return hash(
```
I'm now wondering whether we should maybe still go with `md5` and never regenerate token when `UsersManager.updateUser` is being called. There is not really a benefit of `sha256` over `md5` I'd say as long as the generation is based on random input. 

The `sha256` as discussed in the issue makes more sense when we store the md5 token hashed as sha256. This way the token will be no longer stored as plain text which is a big problem right now. This is a feature for another PR though and will not be in the 3.0 release. Still going with `md5` could allow us later to differentiate between a plain md5 token with 32 char length and a sha256 hashed md5 token (which is 64 length hashed in db). We will likely store auth tokens hashed as part of  https://github.com/piwik/piwik/issues/6559 . When implementing this we will likely use a new table for storing auth tokens (app specific tokens/passwords) anyway and we can also hash all tokens on "update" and write them into a different table eg `user_tokens`. So this differentation between md5 and sha256 will be likely not needed actually and we can keep using sha256 in this PR for token generation 👍  So I presume no change needed, just wanted to write down what came to my mind before :)

mattab replied:
Is there an advantage to making the API token sha256 over md5? one downside which I'm only thinking about now is that changing the `token_auth` length also changes our APIs which may need some changes in SDK APIs, as SDKs tend to test that token_auth is 32 chars or throw an exception otherwise (as we do this in the php sdk). Maybe after all it would make sense to keep md5 for api tokens?

mneudert replied:
It was already changed by @tsteur, but the original intention was to change the length of the hashes to be able to keep the "changing password changes token" in place. It might have been sha1 as well, there was no real argument it had to be sha256.

If I had known the length checks in the SDKs I would have probably chosen something different :+1:

tsteur commented on plugins/UsersManager/templates/index.twig:
```diff
@@ -191,7 +209,13 @@
                         <td id="email" class="editable" ng-click='manageUsers.editUser("row{{ i|e('js') }}")'>{{ user.email }}</td>
                         <td id="alias" class="editable" ng-click='manageUsers.editUser("row{{ i|e('js') }}")'>{{ user.alias|raw }}</td>
-                        <td id="token_auth_user" class="token_auth" data-token="{{ user.token_auth }}">{{ user.token_auth|slice(0, 8) }}…</td>
+                        <td id="token_auth_user" class="token_auth" data-token="{{ user.token_auth }}">
```
@mattab Ideally we remove the tokens of all users here. 

It is off topic and does not belong to this PR and will be done as part of https://github.com/piwik/piwik/issues/6559 anyway at the latest but could be a good opportunity to remove them already in Piwik 3 from UsersManagement list. This is a security problem eg when there are 3 super users, and you can see the tokens of all other super users. Then you leave the company but still know the token of all other super users and users

@mattab maybe we can remove them already now in this PR as we want to improve security

tsteur replied:
We ca also do this in a separate PR for Piwik 3 as we would maybe also need to remove them from API

tsteur replied:
I'd probably even remove the possibility to regenerate the token for any user then. I'm not seeing a big use case for this as one can always log in and do it oneself. It is useful in some rare cases but not sure if this justifies that you could regenerate the token of someone else and the other person doesn't know about it and it may break applications where the other user uses the token

mattab replied:
after you mention it, absolutely agree :+1:  

@mneudert to confirm,  we'd like to do the following:

* Remove the token_auth listed in the UI (when logged in as Super User) -> improves security as the page may be transferred over non secure networks, or may be cached in computers, etc. 
* Remove possibility for super users to regenerate tokens of other users.

tsteur commented on plugins/UsersManager/lang/en.json:
```diff
@@ -63,6 +64,11 @@
         "TheLoginScreen": "The login screen",
         "ThereAreCurrentlyNRegisteredUsers": "There are currently %s registered users.",
+        "TokenAuth": "API Authentication Token",
+        "TokenRegenerateConfirm": "Changing the API authentication token will invalidate the users current token. If the user is currently logged in he may be required to login again. Do you really want to continue?",
```
required to login again => " required to log in again"

tsteur replied:
I'm not sure. Is it "will invalidate the users current token" => "will invalidate the users' current token"?

tsteur replied:
"currently logged in he may be" => "currently logged in, he may be"

mneudert replied:
Let's get some grammar guidance here :D

According to what I understand from http://www.grammarbook.com/punctuation/apostro.asp we have a display of possession between the user and the token. So probably right to have "user's" here as we are only talking about a single user and his token.

(for self reference regarding "logged in, he may": http://www.thepunctuationguide.com/comma.html#complexsentences)

But I think someone with a deeper knowledge and mastery of english punctuation rules will eventually find any errors left :D

tsteur commented on plugins/UsersManager/lang/en.json:
```diff
@@ -63,6 +64,11 @@
         "ThereAreCurrentlyNRegisteredUsers": "There are currently %s registered users.",
+        "TokenAuth": "API Authentication Token",
+        "TokenRegenerateConfirm": "Changing the API authentication token will invalidate the users current token. If the user is currently logged in he may be required to login again. Do you really want to continue?",
+        "TokenRegenerateConfirmSelf": "Changing the API authentication token will invalidate your current token. Proceeding may be required to login again. Do you really want to continue?",
```
Maybe "Changing the API authentication token will invalidate your own token. You may be required to log in again if you proceed. Do you really want to change your authentication token?"

tsteur commented:
@mneudert 👍  well done. Only have 2 minor translation updates I think. Ideally we do not show the auth tokens of all users anymore in the UI but we can do this in a separate PR / issue as it is not related to this PR

mneudert commented on plugins/Login/PasswordResetter.php:
```diff
@@ -383,7 +395,11 @@ protected function getUserInformation($loginOrMail)
+        $hashInfo = $this->passwordHelper->info($passwordHash);
+
+        if (!isset($hashInfo['algo']) || 0 >= $hashInfo['algo']) {
+            throw new Exception(Piwik::translate('Login_ExceptionPasswordMD5HashExpected'));
```
Just saw the screenshot was failing: http://builds-artifacts.piwik.org/piwik/piwik/3.x-dev/21123/Login_password_reset_complete.png

After modifying the \UsersManager\Model to accept only fully hashed passwords I noticed the resetter still stored the new password not fully hashed. However after updating `savePasswordResetInfo()` the old sanity check still expected a 32 char hash.

I changed the sanity check to require a string compatible with `password_hash()`. This should be safe even when the default algorithm changes (sometime far in the future...).

mneudert replied:
Screenshots now look fine. Well, at least the ones I get listed (only those with differences?). The "all screenshots" button throws an exception at me and the tool's repository is private so I cannot get the real path (guest privileges) 🙄

mattab replied:
@mneudert thanks for the report Marc, just fixed this issue so now this button `View all processed screenshots` should work again!

tsteur commented on plugins/UsersManager/lang/en.json:
```diff
@@ -8,7 +8,8 @@
         "ChangeAllConfirm": "Are you sure you want to give '%s' access to all the websites?",
-        "ChangePasswordConfirm": "Changing the password will also change the user's token_auth. Do you really want to continue?",
+        "ChangePasswordConfirm": "Changing the password will also change the user's old API authentication token which may be used to access the Piwik APIs. This token will be re-generated only once. This is a security measure to keep your Piwik data safe. Do you really want to continue?",
+        "ChangePasswordConfirmSelf": "Changing the password will also change the your old API authentication token which may be used to access the Piwik APIs. This token will be re-generated only once. This is a security measure to keep your Piwik data safe. Proceeding may require you to log in again. Do you really want to continue?",
```
"change the your old API authentication"

=> "change your old API authentication" :)

mneudert replied:
There is always that one "typo" creeping in... 😒

tsteur commented:
Looks good to merge to me. One translation change is needed but can do this later. @mattab  ok to merge and then release a beta version?

mattab commented on plugins/UsersManager/lang/en.json:
```diff
@@ -8,7 +8,8 @@
         "ChangeAllConfirm": "Are you sure you want to give '%s' access to all the websites?",
-        "ChangePasswordConfirm": "Changing the password will also change the user's token_auth. Do you really want to continue?",
+        "ChangePasswordConfirm": "Changing the password will also change the user's old API authentication token which may be used to access the Piwik APIs. This token will be re-generated only once. This is a security measure to keep your Piwik data safe. Do you really want to continue?",
+        "ChangePasswordConfirmSelf": "Changing the password will also change your old API authentication token which may be used to access the Piwik APIs. This token will be re-generated only once. This is a security measure to keep your Piwik data safe. Proceeding may require you to log in again. Do you really want to continue?",
```
@mneudert it would be great to refactor translation strings so that each sub-string ideally only appears once. This has many advantages especially around making text consistent in different languages and keeping translators working more efficiently :-) For example here, the string can be `Proceeding may require you to log in again.` and reuse `ChangePasswordConfirm`

mattab commented on plugins/UsersManager/lang/en.json:
```diff
@@ -63,6 +64,11 @@
         "ThereAreCurrentlyNRegisteredUsers": "There are currently %s registered users.",
+        "TokenAuth": "API Authentication Token",
+        "TokenRegenerateConfirm": "Changing the API authentication token will invalidate the user's current token. If the user is currently logged in, he may be required to log in again. Do you really want to continue?",
+        "TokenRegenerateConfirmSelf": "Changing the API authentication token will invalidate your own token. You may be required to log in again if you proceed. Do you really want to change your authentication token?",
```
here you could reuse the string `TokenRegenerateConfirmSelf` within `TokenRegenerateConfirmSelf` (removing the substring ` You may be required to log in again if you proceed.`)

tsteur commented:
@mneudert for now the only change needed would be to move `sha256` back to `md5` so we don't break APIs and SDK. Any chance you can work on this in the next 24 hours? As It is a small change I reckon we could otherwise also merge this PR and I apply the change afterwards

tsteur commented:
Merging @mneudert Kudos !!! so good to have secure passwords. I will make the changes here and switch to md5 quickly
